File: 199BinaryTreeRightSideView/BinaryTreeRightSideView.go
package main

import (
	"fmt"
	nodestructure "leetcode/node_structure"
)

/*
BFS
*/
func rightSideView(root *nodestructure.TreeNode) []int {
	ans := []int{}
	if nil == root {
		return ans
	}
	q := []*nodestructure.TreeNode{root}
	for len(q) > 0 {
		qsize := len(q)
		for i := 0; i < qsize; i++ {
			cur := q[0]
			q = q[1:]
			if nil != cur.Left {
				q = append(q, cur.Left)
			}
			if nil != cur.Right {
				q = append(q, cur.Right)
			}
			if i == qsize-1 {
				ans = append(ans, cur.Val)
			}
		}
	}
	return ans
}

func main() {
	fmt.Println("wait for TreeNode")
}

File: 0199BinaryTreeRightSideView/BinaryTreeRightSideView.py
# -*- coding: utf-8 -*-

import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
from node_structure.TreeNode import TreeNode
from typing import Optional, List

class BinaryTreeRightSideView:
    def rightSideView1(self, root: "Optional[TreeNode]") -> "List[int]":
        """
        bfs, append right most element in each level
        """
        if not root: return []
        ans = []
        q = [root]
        while q:
            qsize = len(q)
            for i in range(qsize):
                cur = q.pop(0)
                if cur.left: q.append(cur.left)
                if cur.right: q.append(cur.right)
                if i == qsize - 1:
                    ans.append(cur.val)
        return ans


def main():
    test = BinaryTreeRightSideView()
    print(test.rightSideView1(
        TreeNode.fromList([1, 2, 3])
    ))

if __name__ == "__main__":
    main()
